src/core/fetchers/cdn_prioritizer.ts

```diff
@@ -0,0 +1,410 @@
+        log.warn("CP: Steering Manifest initially present but no way to fetch it.");
+        this._readyState = createSharedReference("ready");
+      } else {
+        const readyState = manifest.contentSteering.queryBeforeStart ? "not-ready" :
+                                                                       "ready";
+        this._readyState = createSharedReference(readyState);
+        this._autoRefreshSteeringManifest(steeringManifestFetcher,
+                                          manifest.contentSteering);
+      }
+    } else {
+      this._readyState = createSharedReference("ready");
+    }
+    destroySignal.register(() => {
+      this._readyState.setValue("disposed");
+      this._readyState.finish();
+      this._steeringManifestUpdateCanceller?.cancel();
+      this._steeringManifestUpdateCanceller = null;
+      this._lastSteeringManifest = null;
+      for (const timeout of this._downgradedCdnList.timeouts) {
+        clearTimeout(timeout);
+      }
+      this._downgradedCdnList = { metadata: [], timeouts: [] };
+    });
+  }
+
+  /**
+   * From the list of __ALL__ CDNs available to a resource, return them in the
+   * order in which requests should be performed.
+   *
+   * Note: It is VERY important to include all CDN that are able to reach the
+   * wanted resource, even those which will in the end not be used anyway.
+   * If some CDN are not communicated, the `CdnPrioritizer` might wrongly
+   * consider that the current resource don't have any of the CDN prioritized
+   * internally and return other CDN which should have been forbidden if it knew
+   * about the other, non-used, ones.
+   *
+   * @param {Array.<string>} everyCdnForResource - Array of ALL available CDN
+   * able to reach the wanted resource - even those which might not be used in
+   * the end.
+   * @returns {Object} - Array of CDN that can be tried to reach the
+   * resource, sorted by order of CDN preference, according to the
+   * `CdnPrioritizer`'s own list of priorities.
+   *
+   * This value is wrapped in a `ISyncOrAsyncValue` as in relatively rare
+   * scenarios, the order can only be known once the steering Manifest has been
+   * fetched.
+   */
+  public getCdnPreferenceForResource(
+    everyCdnForResource : ICdnMetadata[]
+  ) : ISyncOrAsyncValue<ICdnMetadata[]> {
+    if (everyCdnForResource.length <= 1) {
+      // The huge majority of contents have only one CDN available.
+      // Here, prioritizing make no sense.
+      return SyncOrAsync.createSync(everyCdnForResource);
+    }
+
+    if (this._readyState.getValue() === "not-ready") {
+      const val = new Promise<ICdnMetadata[]>((res, rej) => {
+        this._readyState.onUpdate((readyState) => {
+          if (readyState === "ready") {
+            res(this._innerGetCdnPreferenceForResource(everyCdnForResource));
+          } else if (readyState === "disposed") {
+            rej(new CancellationError());
+          }
+        });
```

Comment:
Is it possible that the ready states goes to `ready`before registering the onUpdate listener ?

Reply:
It shouldn't because we just checked its value (set to "not-ready") and because a callback inside a Promise declaration is called right away.
i.e.:
```
console.log(1);
new Promise(() => {
  console.log(2);
})
console.log(3);
```
Produce the following output:
```
1
2
3
```
_(Only the callback put in a then/catch method - or what's after an `await` keyword - is scheduled in a microtask, where other things could happen in between._

Though you're right that it may not be clear that this is going on, as we're still in a callback and not everyone should know about this complex JS intricacy. In your opinion, should I do something like also ask for the current value, or document?

Reply:
Oh I didn't know about this behavior, but no need for additional precision